src/Microsoft.ComponentDetection.Detectors/pip/IPyPiClient.cs

```diff
@@ -84,54 +92,88 @@ public PyPiClient(IEnvironmentVariableService environmentVariableService, ILogge
 
     public static HttpClient HttpClient { get; internal set; } = new HttpClient(HttpClientHandler);
 
-    public async Task<IList<PipDependencySpecification>> FetchPackageDependenciesAsync(string name, string version, PythonProjectRelease release)
+    private static async Task<HttpResponseMessage> SendGetRequestAsync(
+        Uri uri,
+        HttpCompletionOption completionOption = HttpCompletionOption.ResponseContentRead)
     {
-        var dependencies = new List<PipDependencySpecification>();
+        using var request = new HttpRequestMessage(HttpMethod.Get, uri);
+        request.Headers.UserAgent.Add(ProductValue);
+        request.Headers.UserAgent.Add(CommentValue);
+        var response = await HttpClient.SendAsync(request, completionOption);
+        return response;
+    }
 
+    public async Task<IList<PipDependencySpecification>> FetchPackageDependenciesAsync(string name, string version, PythonProjectRelease release)
+    {
         var uri = release.Url;
-        var response = await this.GetAndCachePyPiResponseAsync(uri);
+        var key = new CacheDependencyKey(name, version, uri);
 
-        if (!response.IsSuccessStatusCode)
+        return (await this.cachedDependencies.GetOrCreateAsync(key, FetchDependencies)).ToList();
```

Comment:
nit: .ToList might be redundant since the task result is already a list and it eventually gets converted to dictionary in pythonresolver.